localstack-core/localstack/services/stepfunctions/asl/component/state/state_execution/state_map/tolerated_failure.py

```diff
@@ -1,44 +1,155 @@
+from localstack.services.stepfunctions.asl.component.common.error_name.failure_event import (
+    FailureEvent,
+    FailureEventException,
+)
+from localstack.services.stepfunctions.asl.component.common.error_name.states_error_name import (
+    StatesErrorName,
+)
+from localstack.services.stepfunctions.asl.component.common.error_name.states_error_name_type import (
+    StatesErrorNameType,
+)
 from localstack.services.stepfunctions.asl.component.eval_component import EvalComponent
 from localstack.services.stepfunctions.asl.eval.environment import Environment
+from localstack.services.stepfunctions.asl.eval.event.event_detail import EventDetails
+from localstack.services.stepfunctions.asl.utils.encoding import to_json_str
 from localstack.services.stepfunctions.asl.utils.json_path import JSONPathUtils
 
+TOLERATED_FAILURE_COUNT_MIN: Final[int] = 0
+TOLERATED_FAILURE_COUNT_DEFAULT: Final[int] = 0
+TOLERATED_FAILURE_PERCENTAGE_MIN: Final[float] = 0.0
+TOLERATED_FAILURE_PERCENTAGE_DEFAULT: Final[float] = 0.0
+TOLERATED_FAILURE_PERCENTAGE_MAX: Final[float] = 100.0
 
-class ToleratedFailureCount(EvalComponent):
-    DEFAULT: Final[int] = 0  # No threshold.
 
-    def __init__(self, count: int = DEFAULT):
-        self.count: Final[int] = count
+class ToleratedFailureCountDecl(EvalComponent, abc.ABC):
+    @abc.abstractmethod
+    def _eval_tolerated_failure_count(self, env: Environment) -> int: ...
 
     def _eval_body(self, env: Environment) -> None:
-        env.stack.append(self.count)
+        tolerated_failure_count = self._eval_tolerated_failure_count(env=env)
+        env.stack.append(tolerated_failure_count)
 
 
-class ToleratedFailureCountPath(EvalComponent):
-    def __init__(self, path: str):
-        self.path: Final[str] = path
+class ToleratedFailureCount(ToleratedFailureCountDecl):
+    tolerated_failure_count: Final[int]
 
-    def _eval_body(self, env: Environment) -> None:
-        value = JSONPathUtils.extract_json(self.path, env.stack[-1])
-        env.stack.append(copy.deepcopy(value))
+    def __init__(self, tolerated_failure_count: int = TOLERATED_FAILURE_COUNT_DEFAULT):
+        self.tolerated_failure_count = tolerated_failure_count
 
+    def _eval_tolerated_failure_count(self, env: Environment) -> int:
+        return self.tolerated_failure_count
 
-class ToleratedFailurePercentage(EvalComponent):
-    DEFAULT: Final[float] = 0.0
 
-    def __init__(self, percentage: float = DEFAULT):
-        self.percentage: Final[float] = percentage
+class ToleratedFailureCountPath(ToleratedFailureCountDecl):
+    tolerated_failure_count_path: Final[str]
 
-    def _eval_body(self, env: Environment) -> None:
-        env.stack.append(self.percentage)
+    def __init__(self, tolerated_failure_count_path: str):
+        self.tolerated_failure_count_path = tolerated_failure_count_path
+
+    def _eval_tolerated_failure_count(self, env: Environment) -> int:
+        inp = env.stack[-1]
```

Comment:
Yes indeed, this will be invoked by a valid parent as specified from the grammar, which means the input is pushed on the stack before evaluation. This is true for all productions in the language